Sources/AWSSDKSwiftCore/AWSClient.swift

```diff
@@ -644,63 +647,89 @@ extension AWSClient {
     }
 
     private func createError(for response: AWSResponse) -> Error {
-        let bodyDict: [String: Any] = (try? response.body.asDictionary()) ?? [:]
+        struct XMLError: Codable, ErrorMessage {
+            var code: String?
+            var message: String
 
-        var code: String?
-        var message: String?
+            private enum CodingKeys: String, CodingKey {
+                case code = "Code"
+                case message = "Message"
+            }
+        }
+        struct QueryError: Codable, ErrorMessage {
+            var code: String?
+            var message: String
+
+            private enum CodingKeys: String, CodingKey {
+                case code = "Code"
+                case message = "Message"
+            }
+        }
+        struct JSONError: Codable, ErrorMessage {
+            var code: String?
+            var message: String
+
+            private enum CodingKeys: String, CodingKey {
+                case code = "__type"
+                case message = "message"
+            }
+        }
+
+        var errorMessage: ErrorMessage? = nil
 
         switch serviceProtocol.type {
         case .query:
             guard case .xml(let element) = response.body else { break }
-            guard let error = element.elements(forName: "Error").first else { break }
-            code = error.elements(forName: "Code").first?.stringValue
-            message = error.elements(forName: "Message").first?.stringValue
+            guard let errorElement = element.elements(forName: "Error").first else { break }
+            errorMessage = try? XMLDecoder().decode(QueryError.self, from: errorElement)
 
         case .restxml:
             guard case .xml(let element) = response.body else { break }
-            code = element.elements(forName: "Code").first?.stringValue
-            message = element.children(of:.element)?.filter({$0.name != "Code"}).map({"\($0.name!): \($0.stringValue!)"}).joined(separator: ", ")
+            errorMessage = try? XMLDecoder().decode(XMLError.self, from: element)
 
         case .restjson:
-            code = response.headers["x-amzn-ErrorType"] as? String
-            message = bodyDict.filter({ $0.key.lowercased() == "message" }).first?.value as? String
+            guard case .json(let data) = response.body else { break }
+            errorMessage = try? JSONDecoder().decode(JSONError.self, from: data)
+            if errorMessage?.code == nil {
+                errorMessage?.code = response.headers["x-amzn-ErrorType"] as? String
+            }
 
         case .json:
-            code = bodyDict["__type"] as? String
-            message = bodyDict.filter({ $0.key.lowercased() == "message" }).first?.value as? String
+            guard case .json(let data) = response.body else { break }
```

Comment:
I'm not sure, if we should handle the error here with the default `unknown error` further down. maybe protocol type mismatch error?

Reply:
The error handling I’ve pretty much inherited. I’ve found the error format the json services return is not totally consistent. I need to investigate this further. I’ll add an issue to look into this later. Would that be enough for now? This PR doesn’t change how aws-sdk-swift responds to errors it just makes the handling code a little cleaner.